xdsl/transforms/printf_to_putchar.py

```diff
@@ -0,0 +1,148 @@
+        func_call = func.Call.get("putchar", [op.char], [i32])
+        # Add empty new_results, since a result is necessary for linking
+        # putchar, but the result does not exist anywhere.
+        rewriter.replace_matched_op(func_call, new_results=[])
+
+
+class ConvertPrintIntToItoa(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_int to an
+    inline_itoa function
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintIntOp, rewriter: PatternRewriter, /):
+        func_call = func.Call.get("inline_itoa", [op.int], [])
+        rewriter.replace_matched_op(func_call)
+
+
+def get_inline_itoa():
+    inline_itoa_func = func.FuncOp("inline_itoa", ((i32,), ()))
+    with ImplicitBuilder(inline_itoa_func.body) as (integer,):
+        zero = arith.Constant.from_int_and_width(0, i32)
+        is_negative = arith.Cmpi.get(integer, zero, "slt")
+        # If the value is zero, just print one zero
+        is_zero = arith.Cmpi.get(integer, zero, "eq")
+        is_zero_block = Block()
+        # If is_zero is true:
+        with ImplicitBuilder(is_zero_block):
+            ascii_zero = arith.Constant.from_int_and_width(48, i32)
+            # The ascii value for zero is 48 in decimal
+            PrintCharOp((ascii_zero,))
+            scf.Yield.get()
+        # If is_zero is false:
+        is_not_zero_block = Block()
+        with ImplicitBuilder(is_not_zero_block):
+            # If the value is negative, print the minus sign, return abs value
+            is_negative_block = Block()
+            with ImplicitBuilder(is_negative_block):
+                ascii_minus = arith.Constant.from_int_and_width(45, i32)
+                PrintCharOp((ascii_minus,))
+                negative_input = arith.Subi(zero, integer)
+                scf.Yield.get(negative_input)
+            # If the value is positive, just continue
+            is_positive_block = Block()
+            with ImplicitBuilder(is_positive_block):
+                scf.Yield.get(integer)
+            absolute_value = scf.If.get(
+                is_negative, [i32], [is_negative_block], [is_positive_block]
+            )
+            one = arith.Constant.from_int_and_width(1, i32)
+            plus_one = arith.Addi(absolute_value, one)
+            # FIXME has to be unsigned
+            plus_one_fp = arith.SIToFPOp((plus_one,), (f32,))
+            size = math.Log10Op((plus_one_fp,), (f32,))
+            size_ceiled = math.CeilOp((size,), (f32,))
+            # FIXME has to be unsigned
+            size_int = arith.FPToSIOp((size_ceiled,), (i32,))
+
+            zero_index = arith.Constant.from_int_and_width(0, IndexType())
+            one_index = arith.Constant.from_int_and_width(1, IndexType())
+            size_int_index = arith.IndexCastOp((size_int,), (IndexType(),))
+
+            loop_body = Block(arg_types=(IndexType(),))
+            with ImplicitBuilder(loop_body) as (index_var,):
```

Comment:
I don't think a lot can be done about this section; it's rather straightforward.
Maybe clean up and provide a clear comment.

If it's worth factoring out (for readability, testing, etc), I think it will depend on what you decide to do in the previous parts, seeing how much they improve this.